### DISCUSSION TITLE
How to put *.egg-info/ directory out of src/ using setup.cfg or pyproject.toml?

### DISCUSSION THREAD

**Discussion by JDLH** [pypa/setuptools#3348] [Q&A]
I am new to **setuptools**. I have started using it to package up a small, 1-module utility for internal use. 

I am surprised to see, in my `src/` directory where my Python modules lives, a `src/My_Project.egg_info/*` directory. It appears to have metadata in it. It appears to get constructed by **setuptools** when it performs the build. 

It doesn't seem sensible to have this build artifact, with project metadata, in my `src/` directory. I would like to locate it in its own directory, as a sibling to `src/`. Maybe put it in a directory `build-info/`. How can I tell **setuptools** where to put this directory as it builds my project?

The only documentation I can find about controlling the construction of this `....egg_info/` directory is in the [*Command Reference* for **setuptools**, section "egg_info - Create egg metadata and set build tags"](https://setuptools.pypa.io/en/latest/userguide/commands.html#egg-info). As far as I can tell, everything in that section refers to commands issued from the shell via `setup.py`. But I am not using `setup.py` now, and would like to avoid it. From what I understand, `setup.py` is deprecated, and I should not use it in this new project.

I am controlling **setuptools** with a `setup.cfg` file, and a minimal `pyproject.toml` file.  How can I tell **setuptools** where to put this `....egg_info/` directory via either `setup.cfg` or `pyproject.toml`?  And where is this documented?

**Comment by abravalheri** [comment]
Hi @JDLH, I have never tried myself (and I don't know how it will play together with other package tools like `build`/`pip`) but you can try specifying a section in the `setup.cfg` file with the same arguments that you would provide for `python setup.py`. For example:

```ini
[egg_info]
egg_base = ...
```

You might already know the following information (so please feel free to ignore the following items):

1. You can consider the `.egg-info` directory as a "cache artifact" in your file system. So you might also add it to your ".gitignore" file (in the same way you might be already ignoring the `__pycache__` directory automatically created by the Python executable)
2. In the future `setuptools` is very likely to get rid of the `.egg-info` directory. We are in a long journey to achieve that and there are some tasks we need to complete first...

**Reply by szabi** [reply]
Not original poster, but

```
[egg_info]
egg_base = build
```

worked like a charm for me!  🎉



I also found an argument, why one might want to keep it outside of a temp/cache/build directory:

If one is putting files of their own into the .egg-info (like i18n resources for [EggTranslations](http://pypi.python.org/pypi/EggTranslations)), then one wants those to be checked in. **Source:** an [S/O answer](https://stackoverflow.com/questions/3575493/do-i-need-egg-info-directories-when-using-setuptools-distribute-to-create-a-py#3623928)

**Reply by stefaneidelloth** [reply]
What is the corresponding entry for pyproject.toml and where to place it?

**Reply by szabi** [reply]
@stefaneidelloth no idea, could not figure it out yet. If you find out, please post it here. I'd love to put it in my `pyproject.toml`.

Currently I have everything in the pyproject file, but this setting is sitting solo in `setup.cfg`.

**Reply by stefaneidelloth** [reply]
I switched to hatchling as build-backend. Also see https://stackoverflow.com/questions/74508024/is-requirements-txt-still-needed-when-using-pyproject-toml and https://github.com/pypa/pip/issues/12100

**Reply by spiltcoffee** [reply]
For anyone looking for the answer to @stefaneidelloth 's question, you want to add this to your `pyproject.toml`:
```
[tool.distutils.egg_info]
egg_base = "target"
```
At least, that's what finally got the directory to move for me.

**Comment by JDLH** [comment]
Thank you for the prompt answer @abravalheri . It is very helpful. I will give it a try.

There are a couple of insights in your answer which I did not see in the documentation. 1. There is a way of transforming the syntax of a `setup.py` command into a directive in `setup.cfg`. If this is documented, it helps people apply the rest of the *Command Reference* to `setup.cfg` cases. 2. The `.egg-info` directory is a cache artifact. I had a hard time finding any documentation about this directory at all. If setuptools makes it, it would be good for the setuptools documentation to say, "we create this directory, and we put it in the src/ directory, and it is a cache artifact".

**Reply by abravalheri** [reply]
Thank you very much for the feedback @JDLH. Indeed the documentation of setuptools is very lacky, so please feel free to contribute if you feel like doing it (since I joined the project I am focusing on trying to implement the evergrowing number of PEPs 😅, so I did not have the time to tackle all the problems in the docs).

This trick with the `setup.cfg` file is something inherited from `distutils`, so you can read more about it in https://docs.python.org/3/distutils/configfile.html (ideally we should migrate these docs to `setuptools` soon).
